src/logger.sh

```diff
@@ -0,0 +1,234 @@
+# Global variables used in this module
+# -------------------------------------------------------------------------------------------------
+yb_fatal_quiet=${yb_fatal_quiet:-false}
+yb_log_quiet=${yb_log_quiet:-false}
+FAIL_ON_WARNING=${FAIL_ON_WARNING:-false}
+
+# -------------------------------------------------------------------------------------------------
+# Global variables defined in this module
+# -------------------------------------------------------------------------------------------------
+# shellcheck disable=SC2034,SC2155
+readonly YB_BASH_COMMON_ROOT=$( cd "${BASH_SOURCE/*}" && cd .. && pwd )
+
+# shellcheck disable=SC2034,SC2155
+readonly YELLOW_COLOR="\033[0;33m"
+
+# shellcheck disable=SC2034,SC2155
+readonly RED_COLOR="\033[0;31m"
+
+# shellcheck disable=SC2034,SC2155
+readonly CYAN_COLOR="\033[0;36m"
+
+# shellcheck disable=SC2034,SC2155
+readonly NO_COLOR="\033[0m"
+
+# shellcheck disable=SC2034,SC2155
+# http://stackoverflow.com/questions/5349718/how-can-i-repeat-a-character-in-bash
+readonly HORIZONTAL_LINE=$( printf '=%.0s' {1..80} )
+
+# -------------------------------------------------------------------------------------------------
+# Colors
+# -------------------------------------------------------------------------------------------------
+
+yellow_color() {
+  echo -ne "$YELLOW_COLOR"
+}
+
+red_color() {
+  echo -ne "$RED_COLOR"
+}
+
+cyan_color() {
+  echo -ne "$CYAN_COLOR"
+}
+
+no_color() {
+  echo -ne "$NO_COLOR"
+}
+
+# -------------------------------------------------------------------------------------------------
+# Timestamps
+# -------------------------------------------------------------------------------------------------
+
+get_timestamp() {
+  date +%Y-%m-%dT%H:%M:%S
+}
+
+get_timestamp_for_filenames() {
+  date +%Y-%m-%dT%H_%M_%S
+}
+
+# -------------------------------------------------------------------------------------------------
+# Logging and stack traces
+# -------------------------------------------------------------------------------------------------
+
+function verbose() {
+  # Print our info messages to stderr and only when asked (YB_VERBOSE=true).
+  local msg="$*"
+  if [[ ${YB_VERBOSE} == "true" ]]; then
+    echo -e "${msg}" >&2
+  fi
+}
+
+print_stack_trace() {
+  local -i i=${1:-1}  # Allow the caller to set the line number to start from.
+  echo "Stack trace:" >&2
+  while [[ $i -lt "${#FUNCNAME[@]}" ]]; do
+    echo "  ${BASH_SOURCE[$i]}:${BASH_LINENO[$((i - 1))]} ${FUNCNAME[$i]}" >&2
+    (( i+=1 ))
+  done
+}
+
+fatal() {
+  if [[ -n "${yb_fatal_quiet:-}" ]]; then
+    yb_log_quiet=$yb_fatal_quiet
+  else
+    yb_log_quiet=false
+  fi
+  yb_log_skip_top_frames=1
+  log "$@"
+  if ! "$yb_log_quiet"; then
+    print_stack_trace 2  # Exclude this line itself from the stack trace (start from 2nd line).
+  fi
+  exit "${yb_fatal_exit_code:-1}"
+}
+
+log_empty_line() {
+  if [[ ${yb_log_quiet:-} == "true" ]]; then
```

Comment:
For variables for which we ensure they are defined at the top of the script, perhaps we don't need to check if they are defined every time we use them.